src/device-state.coffee

```diff
@@ -0,0 +1,512 @@
+
+		@router.post '/v1/reboot', (req, res) =>
+			force = validation.checkTruthy(req.body.force)
+			@deviceState.executeStepAction({ action: 'reboot' }, { force })
+			.then (response) ->
+				res.status(202).json(response)
+			.catch (err) ->
+				if err instanceof updateLock.UpdatesLockedError
+					status = 423
+				else
+					status = 500
+				res.status(status).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.post '/v1/shutdown', (req, res) =>
+			force = validation.checkTruthy(req.body.force)
+			@deviceState.executeStepAction({ action: 'shutdown' }, { force })
+			.then (response) ->
+				res.status(202).json(response)
+			.catch (err) ->
+				if err instanceof updateLock.UpdatesLockedError
+					status = 423
+				else
+					status = 500
+				res.status(status).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.get '/v1/device', (req, res) =>
+			@deviceState.getStatus()
+			.then (state) ->
+				stateToSend = _.pick(state.local, [
+					'api_port'
+					'commit'
+					'ip_address'
+					'status'
+					'download_progress'
+					'os_version'
+					'supervisor_version'
+					'update_pending'
+					'update_failed'
+					'update_downloaded'
+				])
+				res.json(stateToSend)
+			.catch (err) ->
+				res.status(500).json({ Data: '', Error: err?.message or err or 'Unknown error' })
+
+		@router.use(@applications.router)
+
+module.exports = class DeviceState extends EventEmitter
+	constructor: ({ @db, @config, @eventTracker }) ->
+		@logger = new Logger({ @eventTracker })
+		@deviceConfig = new DeviceConfig({ @db, @config, @logger })
+		@applications = new ApplicationManager({ @config, @logger, @db, @eventTracker })
+		@on 'error', (err) ->
+			console.error('Error in deviceState: ', err, err.stack)
+		@_currentVolatile = {}
+		_lock = new Lock()
+		@_writeLock = Promise.promisify(_lock.async.writeLock)
+		@_readLock = Promise.promisify(_lock.async.writeLock)
+		@lastSuccessfulUpdate = null
+		@failedUpdates = 0
+		@stepsInProgress = []
+		@applyInProgress = false
+		@scheduledApply = null
+		@applyContinueScheduled = false
+		@shuttingDown = false
+		@_router = new DeviceStateRouter(this)
+		@router = @_router.router
+		@on 'apply-target-state-end', (err) ->
+			if err?
+				console.log("Apply error #{err}")
+			else
+				console.log('Apply success!')
+		@on 'step-completed', (err) ->
+			if err?
+				console.log("Step completed with error #{err}")
+			else
+				console.log('Step success!')
+		@on 'step-error', (err) ->
+			console.log("Step error #{err}")
+
+		@applications.on('change', @reportCurrentState)
+
+	normaliseLegacy: ({ apps, dependentApps, dependentDevices }) =>
+		legacyTarget = { local: { apps: [], config: {} }, dependent: { apps: [], devices: [] } }
+
+		tryParseObj = (j) ->
+			try
+				JSON.parse(j)
+			catch
+				{}
+		tryParseArray = (j) ->
+			try
+				JSON.parse(j)
+			catch
+				[]
+
+		dependentDevices = tryParseArray(dependentDevices)
```

Comment:
How can failing to parse this not be fatal?

Reply:
Yeah, the idea was that really old supervisors might just not have a dependent devices table, so the dependentDevices we get here might be null, and we'd treat it as empty. In any case this is going away with knex migrations.